pennylane/gradients/spsa_gradient.py

```diff
@@ -0,0 +1,573 @@
+        the ``shots`` argument was passed to the transform:
+
+        >>> shots = (10, 100, 1000)
+        >>> dev = qml.device("default.qubit", wires=2, shots=shots)
+        >>> @qml.qnode(dev)
+        ... def circuit(params):
+        ...     qml.RX(params[0], wires=0)
+        ...     qml.RY(params[1], wires=0)
+        ...     qml.RX(params[2], wires=0)
+        ...     return qml.expval(qml.PauliZ(0)), qml.var(qml.PauliZ(0))
+        >>> params = np.array([0.1, 0.2, 0.3], requires_grad=True)
+        >>> qml.gradients.spsa_grad(circuit, shots=shots, h=1e-2)(params)
+        (((array(0.), array(0.), array(0.)), (array(0.), array(0.), array(0.))),
+         ((array(-1.4), array(-1.4), array(-1.4)),
+          (array(2.548), array(2.548), array(2.548))),
+         ((array(-1.06), array(-1.06), array(-1.06)),
+          (array(1.90588), array(1.90588), array(1.90588))))
+
+        The outermost tuple contains results corresponding to each element of the shot vector,
+        as is also visible by the increasing precision.
+        Note that the stochastic approximation and the fluctuations from the shot noise
+        of the device accumulate, leading to a very coarse-grained estimate for the gradient.
+    """
+    if argnum is None and not tape.trainable_params:
+        return _no_trainable_grad_new(tape, shots)
+
+    if validate_params:
+        if "grad_method" not in tape._par_info[0]:
+            gradient_analysis(tape, grad_fn=_spsa_grad_new)
+        diff_methods = grad_method_validation("numeric", tape)
+    else:
+        diff_methods = ["F" for i in tape.trainable_params]
+
+    if all(g == "0" for g in diff_methods):
+        return _all_zero_grad_new(tape, shots)
+
+    gradient_tapes = []
+    extract_r0 = False
+
+    coeffs, shifts = finite_diff_coeffs(n=n, approx_order=approx_order, strategy=strategy)
+
+    if 0 in shifts:
+        # Finite difference formula includes a term with zero shift.
+
+        if f0 is None:
+            # Ensure that the unshifted tape is appended to the gradient tapes
+            gradient_tapes.append(tape)
+            extract_r0 = True
+
+        # Skip the unshifted tape
+        shifts = shifts[1:]
+
+    method_map = choose_grad_methods(diff_methods, argnum)
+
+    indices = [
+        i for i, _ in enumerate(tape.trainable_params) if (i in method_map and method_map[i] != "0")
+    ]
+
+    tapes_per_grad = len(shifts)
+    all_coeffs = []
+    for idx_rep in range(num_directions):
+        direction = sampler(indices, len(tape.trainable_params), idx_rep)
+        inv_direction = qml.math.divide(
+            1, direction, where=(direction != 0), out=qml.math.zeros_like(direction)
+        )
+        # Use only the non-zero part of `direction` for the shifts, to skip redundant zero shifts
+        _shifts = qml.math.tensordot(h * shifts, direction[indices], axes=0)
+        all_coeffs.append(qml.math.tensordot(coeffs / h**n, inv_direction, axes=0))
```

Comment:
Why do we use the inverse for the main coeff and not directions?

Reply:
Just to make sure: Do you mean why we use `inv_direction` over `direction`?
This can be seen as part of the definition of the SPSA gradient, see e.g. the equation labelled with **SP** in the beginning of [this SPSA Wikipedia entry](https://en.wikipedia.org/wiki/Simultaneous_perturbation_stochastic_approximation). Alternatively, you could consider SPSA to perform finite differences in all directions simultaneously, but with wrong evaluation points for each individual direction. In the gradient recipes for finite differences, we have a structure like
$\frac{1}{h} \sum_{i=1}^n a_i * f(h*x_i)$ for some coefficients $a_i$, evaluation shifts $x_i$ and the shift scale $h$. This division by $h$ in front basically is multiplication by `inv_direction`, just for the one-dimensional case.
I hope this makes sense :)

Reply:
Yes I was talking about `inv_direction`. Thanks for the link and explanation. Just to make sure, are we talking about $J(u_n -c_n \Delta_n)$?

Reply:
(updated my comment)

Reply:
Yes exactly, the $\Delta_n$ is the direction vector (of the $n$-th optimization iteration, in this case).